diff --git a/_oasis b/_oasis
index ff6ebfb..75aec10 100644
--- a/_oasis
+++ b/_oasis
@@ -28,7 +28,7 @@ Library "nbd-lwt"
 Executable "nbd-tool"
   Path: cli
   MainIs: main.ml
-  Install: true
+  Install: false
   BuildDepends: nbd.lwt, cmdliner, lwt, lwt.unix, lwt.syntax
 
 Executable "nbd-test-lwt"
diff --git a/setup.ml b/setup.ml
index 6f92335..f134141 100644
--- a/setup.ml
+++ b/setup.ml
@@ -6828,7 +6828,7 @@ let setup_t =
                   },
                    {
                       bs_build = [(OASISExpr.EBool true, true)];
-                      bs_install = [(OASISExpr.EBool true, true)];
+                      bs_install = [(OASISExpr.EBool true, false)];
                       bs_path = "cli";
                       bs_compiled_object = Byte;
                       bs_build_depends =
